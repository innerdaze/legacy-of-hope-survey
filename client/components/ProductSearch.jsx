import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { map } from 'lodash'
import Search from 'grommet/components/Search'

class ProductSearch extends Component {
  constructor(props) {
    super(props)

    this.handleSearch = this.handleSearch.bind(this)
    this.handleSelect = this.handleSelect.bind(this)
  }

  handleSearch({ target }) {
    if (target.value.length > 2) {
      this.props.search(target.value)
    }
  }

  handleSelect({ target, suggestion }) {
    this.props.onSelect(suggestion.product)
  }

  render() {
    return (
      <Search
        responsive
        placeHolder='Enter product name'
        dropAlign={{ left: 'left' }}
        onDOMChange={this.handleSearch}
        onSelect={this.handleSelect}
        suggestions={map(
          this.props.results, product => ({
            product,
            label: product.ProductName
          })
        )}
        />
    )
  }
}

ProductSearch.propTypes = {
  search: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  results: PropTypes.array.isRequired
}

PropTypes.defaultProps = {
  search: Function.prototype,
  onSelect: Function.prototype,
  results: []
}

export default ProductSearch
