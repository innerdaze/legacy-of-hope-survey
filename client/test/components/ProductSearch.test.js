import React from 'react'
import ProductSearch from '../../components/ProductSearch'

let wrapper, search, onSelect, results

describe('Testing on ProductSearch...', () => {
  beforeEach(() => {
    search = jest.fn()
    onSelect = jest.fn()
    results = []
    wrapper = shallow(
      <ProductSearch search={search} onSelect={onSelect} results={results} />
    )
  })

  describe('ProductSearch layout', () => {
    test('Expect ProductSearch to have a Search comp', () => {
      expect(wrapper.find('Search')).toHaveLength(1)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test on prop', () => {
    test('Expect search prop function to be called if Search onDOMChange is triggered', () => {
      wrapper.find('Search').prop('onDOMChange')({
        target: {
          value: 'foo'
        }
      })

      expect(search).toBeCalled()
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect search prop function not called if text length less than 2', () => {
      wrapper.find('Search').prop('onDOMChange')({
        target: {
          value: 'fo'
        }
      })

      expect(search).not.toBeCalled()
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect onSelect prop function to be call if Search onSelect is triggered', () => {
      results = [
        ...results,
        {
          product: 'FooProduct',
          ProductName: 'BarName'
        }
      ]
      wrapper.find('Search').prop('onSelect')({
        suggestion: {
          product: 'FooProduct'
        }
      })

      expect(onSelect).toBeCalled()
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect result prop array is mapped before passed to Search suggestions prop', () => {
      results = [
        ...results,
        {
          product: 'FooProduct',
          ProductName: 'BarName'
        }
      ]

      wrapper = shallow(
        <ProductSearch search={search} onSelect={onSelect} results={results} />
      )
      let remappedResult = wrapper.find('Search').prop('suggestions')
      expect(remappedResult[0]).toHaveProperty('product')
      expect(remappedResult[0]).toHaveProperty('label')

      expect(wrapper).toMatchSnapshot()
    })
  })
})
