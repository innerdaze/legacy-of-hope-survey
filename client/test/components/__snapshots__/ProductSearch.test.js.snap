// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on ProductSearch... ProductSearch layout Expect ProductSearch to have a Search comp 1`] = `
<Search
  align="left"
  dropAlign={
    Object {
      "left": "left",
    }
  }
  iconAlign="end"
  inline={false}
  onDOMChange={[Function]}
  onSelect={[Function]}
  placeHolder="Enter product name"
  responsive={true}
  suggestions={Array []}
/>
`;

exports[`Testing on ProductSearch... Test on prop Expect onSelect prop function to be call if Search onSelect is triggered 1`] = `
<Search
  align="left"
  dropAlign={
    Object {
      "left": "left",
    }
  }
  iconAlign="end"
  inline={false}
  onDOMChange={[Function]}
  onSelect={[Function]}
  placeHolder="Enter product name"
  responsive={true}
  suggestions={Array []}
/>
`;

exports[`Testing on ProductSearch... Test on prop Expect result prop array is mapped before passed to Search suggestions prop 1`] = `
<Search
  align="left"
  dropAlign={
    Object {
      "left": "left",
    }
  }
  iconAlign="end"
  inline={false}
  onDOMChange={[Function]}
  onSelect={[Function]}
  placeHolder="Enter product name"
  responsive={true}
  suggestions={
    Array [
      Object {
        "label": "BarName",
        "product": Object {
          "ProductName": "BarName",
          "product": "FooProduct",
        },
      },
    ]
  }
/>
`;

exports[`Testing on ProductSearch... Test on prop Expect search prop function not called if text length less than 2 1`] = `
<Search
  align="left"
  dropAlign={
    Object {
      "left": "left",
    }
  }
  iconAlign="end"
  inline={false}
  onDOMChange={[Function]}
  onSelect={[Function]}
  placeHolder="Enter product name"
  responsive={true}
  suggestions={Array []}
/>
`;

exports[`Testing on ProductSearch... Test on prop Expect search prop function to be called if Search onDOMChange is triggered 1`] = `
<Search
  align="left"
  dropAlign={
    Object {
      "left": "left",
    }
  }
  iconAlign="end"
  inline={false}
  onDOMChange={[Function]}
  onSelect={[Function]}
  placeHolder="Enter product name"
  responsive={true}
  suggestions={Array []}
/>
`;
